qa/setup_packages.py

```diff
@@ -16,34 +16,433 @@
+            Maximum python version supported by this package. If empty there is no upper bound
+        `python_min_ver`: str, optional, default = None
+            Mimimum python version supported by this package. If empty there is no lower bound
+    """
+    def __init__(self, ver, python_min_ver=None, python_max_ver=None):
+        self.ver = ver
+        self.python_min_ver = python_min_ver
+        self.python_max_ver = python_max_ver
+
+    def __bool__(self):
+        return (not self.python_min_ver or parse(PYTHON_VERSION) >= parse(self.python_min_ver)) and \
+           (not self.python_max_ver or parse(PYTHON_VERSION) <= parse(self.python_max_ver))
+
+    def __repr__(self):
+        if self:
+            return self.ver
+        else:
+            return ""
+
+class BasePackage():
+    """Class describing basic methods that package should provide
+
+        Parameters
+        ----------
+        `key`: str
+            Name this package should be queried for
+        `versions`: str
+            List of versions this package is available for
+        `name`: str, , optional, default = None
+            Name of this package used during installation. If empty it is the same as the key
+    """
+    def __init__(self, key, versions, name=None):
+        self.key = key
+        if not name:
+            name = key
+        self.name = name
+        self.versions = versions
+
+    """Clams index to range 0 - num_of_packages - 1
+
+        Parameters
+        ----------
+        `key`: idx: int
+            Index to clamp
+        `cuda_version`: str, optional, default = None
+            Cuda version used for a given index
+    """
+    def clamp_index(self, idx, cuda_version=None):
+        if idx < 0 or idx >= self.get_num_of_version(cuda_version):
+            idx = 0
+        return idx
+
+    """Retrives package name.
+
+        Parameters
+        ----------
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_name(self, cuda_version=None):
+        raise NotImplementedError
+
+    """Retrieves only compatible versions of this package from provided `versions` list
+
+        Parameters
+        ----------
+        `versions`: list
+            List of versions to be checked. All versions that evaluate to True are returned
+    """
+    def filter_versions(self, versions):
+        # convert PckgVer to string
+        return [str(v) for v in versions if v]
+
+    """Get versions at a given index, compatible with provided cuda_version
+
+        Parameters
+        ----------
+        `idx`: int
+            Index of version to retrive. If index is beyond 0-num_ov_versions-1 range
+            it is clamped to it
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_version(self, idx, cuda_version=None):
+        idx = self.clamp_index(idx, cuda_version)
```

Comment:
I'm not sure if clamping logic is the best solution here, but didn't read the whole thing yet.

Reply:
Fo negative value this script is no op already, but if you are asking for the version which index is out of bounds `clamp` is the best thing, next to throw, to do I think.